File: android/app/src/main/java/com/ebike/rpar/geometry/Transforms.kt
package com.ebike.rpar.geometry

import com.ebike.rpar.model.Intrinsics
import com.ebike.rpar.model.MountProfile
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt
import kotlin.math.tan

/**
 * Coordinate frames (SYNC-005 / GEO-006):
 * - device: Android sensor frame
 * - camera: x right, y down, z forward (optical)
 * - vehicle: x right, y forward, z up (centerline)
 * - road: vehicle XY on Z=0, with left-handlebar lateral offset applied
 */
object Transforms {
    fun deg2rad(d: Double) = d * Math.PI / 180.0

    fun rotX(a: Double): DoubleArray {
        val c = cos(a); val s = sin(a)
        return doubleArrayOf(1.0, 0.0, 0.0, 0.0, c, -s, 0.0, s, c)
    }

    fun rotY(a: Double): DoubleArray {
        val c = cos(a); val s = sin(a)
        return doubleArrayOf(c, 0.0, s, 0.0, 1.0, 0.0, -s, 0.0, c)
    }

    fun rotZ(a: Double): DoubleArray {
        val c = cos(a); val s = sin(a)
        return doubleArrayOf(c, -s, 0.0, s, c, 0.0, 0.0, 0.0, 1.0)
    }

    fun mul3(a: DoubleArray, b: DoubleArray): DoubleArray {
        val o = DoubleArray(9)
        for (r in 0..2) for (c in 0..2) {
            o[r * 3 + c] = a[r * 3] * b[c] + a[r * 3 + 1] * b[3 + c] + a[r * 3 + 2] * b[6 + c]
        }
        return o
    }

    fun mulVec(m: DoubleArray, v: DoubleArray): DoubleArray = doubleArrayOf(
        m[0] * v[0] + m[1] * v[1] + m[2] * v[2],
        m[3] * v[0] + m[4] * v[1] + m[5] * v[2],
        m[6] * v[0] + m[7] * v[1] + m[8] * v[2],
    )

    fun transpose(m: DoubleArray) = doubleArrayOf(
        m[0], m[3], m[6], m[1], m[4], m[7], m[2], m[5], m[8],
    )

    fun quatToRot(q: DoubleArray): DoubleArray {
        var x = q[0]; var y = q[1]; var z = q[2]; var w = q[3]
        val n = sqrt(x * x + y * y + z * z + w * w).coerceAtLeast(1e-9)
        x /= n; y /= n; z /= n; w /= n
        return doubleArrayOf(
            1 - 2 * (y * y + z * z), 2 * (x * y - z * w), 2 * (x * z + y * w),
            2 * (x * y + z * w), 1 - 2 * (x * x + z * z), 2 * (y * z - x * w),
            2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x * x + y * y),
        )
    }

    fun cameraFromVehicle(mount: MountProfile): DoubleArray {
        val pitch = deg2rad(mount.pitchDeg)
        val roll = deg2rad(mount.rollDeg)
        val yaw = deg2rad(mount.yawDeg + mount.handlebarNeutralYawDeg)
        val basis = doubleArrayOf(
            1.0, 0.0, 0.0,
            0.0, 0.0, -1.0,
            0.0, 1.0, 0.0,
        )
        return mul3(rotZ(roll), mul3(rotX(pitch), mul3(rotY(yaw), basis)))
    }

    fun cameraTranslationVehicle(mount: MountProfile): DoubleArray =
        doubleArrayOf(mount.lateralOffsetM, 0.0, mount.cameraHeightM)

    fun projectVehiclePoint(pVehicle: DoubleArray, mount: MountProfile, k: Intrinsics): DoubleArray? {
        val r = cameraFromVehicle(mount)
        val tCam = mulVec(r, cameraTranslationVehicle(mount)).map { -it }.toDoubleArray()
        val pCam = DoubleArray(3) { i ->
            r[i * 3] * pVehicle[0] + r[i * 3 + 1] * pVehicle[1] + r[i * 3 + 2] * pVehicle[2] + tCam[i]
        }
        if (pCam[2] <= 0.15) return null
        val u = k.fx * pCam[0] / pCam[2] + k.cx
        val v = k.fy * pCam[1] / pCam[2] + k.cy
        return doubleArrayOf(u, v)
    }

    fun pixelToGround(uv: DoubleArray, mount: MountProfile, k: Intrinsics): DoubleArray? {
        val x = (uv[0] - k.cx) / k.fx
        val y = (uv[1] - k.cy) / k.fy
        var ray = doubleArrayOf(x, y, 1.0)
        val n = sqrt(ray[0] * ray[0] + ray[1] * ray[1] + ray[2] * ray[2])
        if (n < 1e-9) return null
        ray = doubleArrayOf(ray[0] / n, ray[1] / n, ray[2] / n)
        val rVc = cameraFromVehicle(mount)
        val tV = cameraTranslationVehicle(mount)
        val dirV = mulVec(transpose(rVc), ray)
        if (kotlin.math.abs(dirV[2]) < 1e-8) return null
        val scale = -tV[2] / dirV[2]
        if (scale < 0.2) return null
        val p = doubleArrayOf(tV[0] + scale * dirV[0], tV[1] + scale * dirV[1], tV[2] + scale * dirV[2])
        if (!p.all { it.isFinite() } || p[1] < 0.4 || p[1] > 90) return null
        return doubleArrayOf(p[0], p[1])
    }

    fun compensateLeftHandlebar(xyCamGround: DoubleArray, mount: MountProfile): DoubleArray =
        doubleArrayOf(xyCamGround[0] - mount.lateralOffsetM, xyCamGround[1])

    fun overlayMaxErrorPx(mount: MountProfile, k: Intrinsics, n: Int = 5): Double {
        var maxErr = 0.0
        val xs = DoubleArray(n) { i -> -1.6 + 3.2 * i / (n - 1).coerceAtLeast(1) }
        val ys = DoubleArray(n) { i -> 5.0 + 23.0 * i / (n - 1).coerceAtLeast(1) }
        for (x in xs) for (y in ys) {
            val uv = projectVehiclePoint(doubleArrayOf(x, y, 0.0), mount, k) ?: continue
            val xy = pixelToGround(uv, mount, k) ?: continue
            val uv2 = projectVehiclePoint(doubleArrayOf(xy[0], xy[1], 0.0), mount, k) ?: continue
            val dx = uv[0] - uv2[0]
            val dy = uv[1] - uv2[1]
            val e = sqrt(dx * dx + dy * dy)
            if (e > maxErr) maxErr = e
        }
        return maxErr
    }

    fun projectedHorizonY(mount: MountProfile, k: Intrinsics): Double? {
        val uv = projectVehiclePoint(doubleArrayOf(0.0, 80.0, 0.0), mount, k) ?: return null
        return uv[1]
    }

    fun fitMountFromDistanceMarkers(
        mount: MountProfile,
        k: Intrinsics,
        markers: List<Pair<Double, Double>>,
    ): MountProfile {
        if (markers.size < 2) return mount
        var bestPitch = mount.pitchDeg
        var bestHeight = mount.cameraHeightM
        var bestErr = Double.POSITIVE_INFINITY
        var pitch = mount.pitchDeg - 8.0
        while (pitch <= mount.pitchDeg + 8.0 + 1e-6) {
            var height = maxOf(0.55, mount.cameraHeightM - 0.45)
            val hMax = mount.cameraHeightM + 0.45
            while (height <= hMax + 1e-6) {
                val trial = mount.copy(pitchDeg = pitch, cameraHeightM = height)
                var err = 0.0
                var ok = true
                for ((distM, yPx) in markers) {
                    val uv = projectVehiclePoint(doubleArrayOf(0.0, distM, 0.0), trial, k)
                    if (uv == null) {
                        ok = false
                        break
                    }
                    val d = uv[1] - yPx
                    err += d * d
                }
                if (ok && err < bestErr) {
                    bestErr = err
                    bestPitch = pitch
                    bestHeight = height
                }
                height += 0.05
            }
            pitch += 0.5
        }
        return mount.copy(pitchDeg = bestPitch, cameraHeightM = bestHeight, valid = true)
    }

    fun defaultIntrinsics(width: Int = 1920, height: Int = 1080, hfovDeg: Double = 68.0): Intrinsics {
        val fx = (width / 2.0) / tan(deg2rad(hfovDeg) / 2.0)
        return Intrinsics(fx, fx, width / 2.0, height / 2.0, width, height, true)
    }

    fun defaultMount(width: Int = 1920, height: Int = 1080): MountProfile {
        val k = defaultIntrinsics(width, height)
        return MountProfile(
            profileId = "left_handlebar_v1",
            name = "Left handlebar landscape 1x main",
            cameraId = "rear_main",
            landscape = true,
            cameraHeightM = 1.12,
            pitchDeg = 18.0,
            rollDeg = 0.0,
            yawDeg = 0.0,
            lateralOffsetM = -0.32,
            handlebarNeutralYawDeg = 0.0,
            nearReferenceM = 3.0,
            horizonYPx = k.cy * 0.42,
            vehicleCenterlineXPx = width * 0.52,
            valid = true,
        )
    }

    fun slerp(q0: DoubleArray, q1in: DoubleArray, u: Double): DoubleArray {
        var q1 = q1in.copyOf()
        var dot = q0[0] * q1[0] + q0[1] * q1[1] + q0[2] * q1[2] + q0[3] * q1[3]
        if (dot < 0) {
            q1 = q1.map { -it }.toDoubleArray(); dot = -dot
        }
        if (dot > 0.9995) {
            val o = DoubleArray(4) { q0[it] + u * (q1[it] - q0[it]) }
            val n = sqrt(o.sumOf { it * it }).coerceAtLeast(1e-9)
            return o.map { it / n }.toDoubleArray()
        }
        val theta = kotlin.math.acos(dot.coerceIn(-1.0, 1.0))
        val so = sin(theta)
        val a = sin((1 - u) * theta) / so
        val b = sin(u * theta) / so
        val o = DoubleArray(4) { a * q0[it] + b * q1[it] }
        val n = sqrt(o.sumOf { it * it }).coerceAtLeast(1e-9)
        return o.map { it / n }.toDoubleArray()
    }
}

File: python/src/rpar/transforms.py
"""Device / camera / vehicle / road / gravity transforms (SYNC-005, GEO-006)."""

from __future__ import annotations

from dataclasses import replace

import numpy as np

from rpar.models import Intrinsics, MountProfile


def deg2rad(d: float) -> float:
    return d * np.pi / 180.0


def rot_x(a: float) -> np.ndarray:
    c, s = np.cos(a), np.sin(a)
    return np.array([[1, 0, 0], [0, c, -s], [0, s, c]], dtype=np.float64)


def rot_y(a: float) -> np.ndarray:
    c, s = np.cos(a), np.sin(a)
    return np.array([[c, 0, s], [0, 1, 0], [-s, 0, c]], dtype=np.float64)


def rot_z(a: float) -> np.ndarray:
    c, s = np.cos(a), np.sin(a)
    return np.array([[c, -s, 0], [s, c, 0], [0, 0, 1]], dtype=np.float64)


def quat_to_rot(q_xyzw: np.ndarray) -> np.ndarray:
    x, y, z, w = q_xyzw
    n = np.sqrt(x * x + y * y + z * z + w * w)
    x, y, z, w = x / n, y / n, z / n, w / n
    return np.array(
        [
            [1 - 2 * (y * y + z * z), 2 * (x * y - z * w), 2 * (x * z + y * w)],
            [2 * (x * y + z * w), 1 - 2 * (x * x + z * z), 2 * (y * z - x * w)],
            [2 * (x * z - y * w), 2 * (y * z + x * w), 1 - 2 * (x * x + y * y)],
        ],
        dtype=np.float64,
    )


def camera_from_vehicle(mount: MountProfile) -> np.ndarray:
    """Rotation taking vehicle coordinates (x right, y forward, z up) into camera (x right, y down, z forward)."""
    pitch = deg2rad(mount.pitch_deg)
    roll = deg2rad(mount.roll_deg)
    yaw = deg2rad(mount.yaw_deg + mount.handlebar_neutral_yaw_deg)
    # vehicle -> camera optical: look along +Y vehicle, then apply mount errors
    basis = np.array(
        [
            [1.0, 0.0, 0.0],  # camera x = vehicle x (right)
            [0.0, 0.0, -1.0],  # camera y = down = -vehicle z
            [0.0, 1.0, 0.0],  # camera z = forward = vehicle y
        ],
        dtype=np.float64,
    )
    return rot_z(roll) @ rot_x(pitch) @ rot_y(yaw) @ basis


def camera_translation_vehicle(mount: MountProfile) -> np.ndarray:
    # vehicle: x right, y forward, z up. Left handlebar is negative x.
    return np.array([mount.lateral_offset_m, 0.0, mount.camera_height_m], dtype=np.float64)


def project_vehicle_point(p_vehicle: np.ndarray, mount: MountProfile, k: Intrinsics) -> np.ndarray | None:
    r = camera_from_vehicle(mount)
    t = -r @ camera_translation_vehicle(mount)
    p_cam = r @ p_vehicle + t
    if p_cam[2] <= 0.15:
        return None
    uv = k.matrix() @ p_cam
    return np.array([uv[0] / uv[2], uv[1] / uv[2]], dtype=np.float64)


def pixel_to_ground(uv: np.ndarray, mount: MountProfile, k: Intrinsics) -> np.ndarray | None:
    """Ray-plane intersection on vehicle Z=0. Result is already vehicle-centerline XY (GEO-006)."""
    x = (float(uv[0]) - k.cx) / k.fx
    y = (float(uv[1]) - k.cy) / k.fy
    ray_cam = np.array([x, y, 1.0], dtype=np.float64)
    n = np.linalg.norm(ray_cam)
    if n < 1e-9:
        return None
    ray_cam = ray_cam / n
    r_v_to_c = camera_from_vehicle(mount)
    t_v = camera_translation_vehicle(mount)
    dir_v = r_v_to_c.T @ ray_cam
    if abs(dir_v[2]) < 1e-8:
        return None
    scale = -t_v[2] / dir_v[2]
    if scale < 0.2:
        return None
    p = t_v + scale * dir_v
    if not np.isfinite(p).all() or p[1] < 0.4 or p[1] > 90:
        return None
    return p[:2]


def ground_to_pixel(xy: np.ndarray, mount: MountProfile, k: Intrinsics) -> np.ndarray | None:
    p = np.array([xy[0], xy[1], 0.0], dtype=np.float64)
    return project_vehicle_point(p, mount, k)


def compensate_left_handlebar(xy_cam_ground: np.ndarray, mount: MountProfile) -> np.ndarray:
    """Shift from camera-centered ground coords into vehicle-centerline road coords (GEO-006)."""
    return np.array([xy_cam_ground[0] - mount.lateral_offset_m, xy_cam_ground[1]], dtype=np.float64)


def rotate_points_about_principal(points: np.ndarray, cx: float, cy: float, roll_rad: float) -> np.ndarray:
    c, s = np.cos(roll_rad), np.sin(roll_rad)
    r = np.array([[c, -s], [s, c]], dtype=np.float64)
    centered = points - np.array([cx, cy])
    return centered @ r.T + np.array([cx, cy])


def default_intrinsics(width: int = 1920, height: int = 1080, hfov_deg: float = 68.0) -> Intrinsics:
    fx = (width / 2.0) / np.tan(deg2rad(hfov_deg) / 2.0)
    fy = fx
    return Intrinsics(fx=fx, fy=fy, cx=width / 2.0, cy=height / 2.0, width=width, height=height)


def chessboard_overlay_error_px(
    mount: MountProfile | None = None,
    k: Intrinsics | None = None,
    n: int = 5,
) -> float:
    """Numerical self-consistency of the ground <-> pixel chain on a grid (CAM-009 desktop half).

    This only checks that projection and back-projection agree; the preview/AR/model
    coordinate agreement on the phone is measured by `CameraTransformChain` on device.
    """
    mount = mount or default_mount()
    k = k or default_intrinsics()
    errs: list[float] = []
    for x in np.linspace(-1.6, 1.6, n):
        for y in np.linspace(5.0, 28.0, n):
            uv = ground_to_pixel(np.array([x, y]), mount, k)
            if uv is None:
                continue
            xy = pixel_to_ground(uv, mount, k)
            if xy is None:
                continue
            uv2 = ground_to_pixel(xy, mount, k)
            if uv2 is None:
                continue
            errs.append(float(np.linalg.norm(uv - uv2)))
    return max(errs) if errs else 999.0


def display_compensate(
    poly: list[tuple[float, float]],
    yaw_rate: float,
    latency_ms: float,
    frame_w: int,
    focal_px: float | None = None,
) -> list[tuple[float, float]]:
    """GEO-007: shift overlay by predicted camera yaw over display latency.

    dx = f * yaw_rate * latency; without intrinsics assume ~85 deg HFOV (f ~ 0.55 w).
    """
    f = float(focal_px) if focal_px and focal_px > 0 else float(frame_w) * 0.55
    dx = float(yaw_rate) * (latency_ms / 1000.0) * f
    if abs(dx) < 0.5:
        return poly
    return [(x + dx, y) for x, y in poly]


def default_mount(width: int = 1920, height: int = 1080) -> MountProfile:
    """Nominal left-handlebar profile. The stored horizon is the one this pose projects to."""
    k = default_intrinsics(width, height)
    mount = MountProfile(
        profile_id="left_handlebar_v1",
        name="Left handlebar landscape 1x main",
        camera_id="rear_main",
        landscape=True,
        camera_height_m=1.12,
        pitch_deg=18.0,
        roll_deg=0.0,
        yaw_deg=0.0,
        lateral_offset_m=-0.32,
        handlebar_neutral_yaw_deg=0.0,
        near_reference_m=3.0,
        horizon_y_px=0.0,
        vehicle_centerline_x_px=float(width) * 0.52,
        valid=True,
    )
    far = project_vehicle_point(np.array([0.0, 80.0, 0.0]), mount, k)
    horizon = float(far[1]) if far is not None else float(k.cy) * 0.42
    return replace(mount, horizon_y_px=horizon)
